WordPress/gutenberg
# Style engine: move backend scripts to package

Requires: 

- https://github.com/WordPress/gutenberg/pull/39446

## What?

A very naïve attempt at:

- copying package PHP files over to /build
- ensuring `_gutenberg/_wp` renaming is minimized for core patches

## Why?

1. The Style Engine backend is an implementation of the JS version. Keeping them together in the package makes the relationship more conspicuous.
2. The Style Engine is something that might be used across Gutenberg PHP files in the future. The intention is to make backporting to core easier. See discussion over at https://github.com/WordPress/gutenberg/issues/33241#issuecomment-876218087

Historical context/discussion:
- https://github.com/WordPress/gutenberg/issues/33241#issuecomment-876218087
- https://github.com/WordPress/gutenberg/pull/39446#issuecomment-1076050281
- https://github.com/WordPress/gutenberg/pull/39446#discussion_r834186170

## How?
- Move Style Engine class to `packages/style-engine`
- Update Webpack packages script to:
    - copy the class to build
    - rename classes and methods so that they don't conflict with core (`_gutenberg`)

## Testing Instructions
Build the project via `npm run dev` or `npm run build`

Check that the PHP file in `packages/style-engine` is copied to `/build` and:
- renamed to `*-gutenberg.php`
- edited so that class names have `_Gutenberg` appended

Follow the test instructions in https://github.com/WordPress/gutenberg/pull/39446, and check that the margin and padding block support styles are output correctly on the frontend.

### Unit tests

```bash
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/packages/style-engine/phpunit/class-wp-style-engine-test.php
```

```bash
npm run build
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/block-supports/spacing-test.php
